apps/src/applab/AppLabView.jsx

```diff
@@ -34,10 +47,18 @@ var AppLabView = React.createClass({
         <ProtectedStatefulDiv
             id="visualizationResizeBar"
             className="fa fa-ellipsis-v" />
-        <InstructionsWithWorkspace/>
+        <InstructionsWithWorkspace>
+          <CodeWorkspace style={{display: codeWorkspaceVisible ? 'block' : 'none' }}/>
+          {this.props.hasDesignMode && <ProtectedDesignWorkspace/>}
+          {this.props.hasDataMode && <DataWorkspace/>}
+        </InstructionsWithWorkspace>
       </StudioAppWrapper>
     );
   }
 });
 
-module.exports = AppLabView;
+module.exports = connect(state => ({
+  hasDataMode: state.pageConstants.hasDataMode || false,
+  hasDesignMode: state.pageConstants.hasDesignMode || false,
```

Comment:
could just be `!!state.pageConstants.hasDesignMode` (instead of ORing with false)